### DISCUSSION TITLE
Proxying to slow backend services

### DISCUSSION THREAD

**Discussion by mnutt** [puma/puma#3526] [Q&A]
We recently switched our rails app from unicorn to puma. One of the things I've been interested in is how puma can sit in front of all of our backend services and do user auth, permissions checking, request rewriting, etc. I'm somewhat familiar with alternatives [0][1] but I wanted to explore how puma might be able to handle this directly.

Some of these backend services take multiple seconds to respond, or send/receive large files. With unicorn, a user initiating a handful of backend requests can easily tie up all of the workers, denying service to other users. I experimented with increasing puma threads, but saw a degradation in performance on regular non-proxied requests in production. I need to spend more time investigating this, but am hypothesizing it is related to context switching.

So anyway my thought was: can I thread _just_ the request proxying?
* request comes in, check the user's session, fetch db records, and run regular auth hooks
* hijack the rack request
* `Thread.new` to let puma finish and go off to serve other requests
  * make `net/http` origin request to the backend
  * receive origin response, pipe it to the hijacked socket

Here's the initial proof of concept for this: https://github.com/mnutt/rails-threaded-proxy

This seems to work better than I expected it would. `IO.copy_stream` is able to stream a 10GB file out at 100+MB/s with practically no CPU or memory overhead. It seems to be able to support lots of concurrent proxied requests with only a single puma thread. I'm posting here because I'm wondering if this has been tried before and/or what I might be missing that might come back to bite me later. Whether it falls into the "difficult to get right" bucket, or "fundamentally flawed". Some caveats I'm aware of:

* It involves low-level http processing that might be brittle
* I punted on keepalive and just `connection: close` everything. I suspect a connection pool on the origin side might be possible.
* Threading issues:
  * People adding callbacks and potentially not understanding that they run in a separate thread
  * System resource exhaustion from too many threads (I should probably use a thread pool)
  * Puma worker thread getting reaped: my app has the threads-per-worker fixed; I imagine if the parent thread dies it would kill the child. Probably also an issue on puma shutdown?
* I've tested this on linux and mac with good results but am not sure what may or may not work on windows

[0] - Alternative: rails generating presigned tokens and letting the client connect to the backend service directly. We use this today but it varies based on the backend service. Some of these services I'd prefer we didn't expose to third parties even with the presigned tokens.
[1] - Alternative: nginx `auth_request` or similar. I haven't tried it; it would mean that we'd have to add nginx to our dev stack.

**Comment by dentarg** [comment]
This sounds pretty crazy to me 😛 

Have you looked at https://github.com/socketry/falcon or https://github.com/boazsegev/iodine? Perhaps they could serve your use-case better.

If you want to look at other servers doing rack hijack there's https://github.com/rails/rails/tree/main/actioncable and https://github.com/discourse/message_bus 

Maybe @ioquatix @boazsegev or @trevorturk are willing to share some advice or insights on this topic

**Reply by ioquatix** [reply]
People use falcon for proxying and have great results. For the best results you'd want to use `async-http` as a client, as you can directly forward the incoming request and response (streaming).

**Reply by boazsegev** [reply]
If to offer my own 2¢ on the matter...:

As far as I can understand you are looking at a "long-polling" situation, where you want to return a response only after your app collected information from a slow backend – and you want to do this without the blocking the app server.

This, I believe to be contradictory to [Rack's](https://github.com/rack/rack) design pattern and could be considered an anti-pattern where web applications are concerned.

I see a few additional possible solutions which you did not seem to mention in your post:

1. **use a "green threads" server**, such as falcon, as suggested above.

   This should minimize the effects that long-polling has on your Rails app.

1. **avoid long-polling and use push instead, leveraging an alternative protocol, such as WebSockets or SSE**.

    This should allow you to offer an improved user experience, for example, by using cached data as an initial response and later updating this data as more information arrives from the backend(s).

1. **switch from [Rack](https://github.com/rack/rack) (which is based on CGI) to a callback pattern**.

   This would allow you to implement long polling as as "first class citizen" in your app (of which I am personally not a fan), as well as streaming responses (which I am more in favor of).

[The `iodine` gem](https://github.com/boazsegev/iodine) can be used to implement performant solutions using both [WebSockets / SSE](https://github.com/boazsegev/iodine?tab=readme-ov-file#special-http-upgrade-and-sse-support) and Callback approaches (long-polling / streaming).

The [WebSocket / SSE](https://github.com/boazsegev/iodine?tab=readme-ov-file#special-http-upgrade-and-sse-support) solution can be implemented using the current gem version of `iodine` (0.7.x), while the callback long-polling / streaming approach requires the new (as of yet unreleased) [NeoRack](https://github.com/boazsegev/neorack) compliant version (0.8.x).

I hope my observations / suggestions help.

P.S.,

As I am the author of [`iodine`](https://github.com/boazsegev/iodine) and as I am currently working on the new gem version as well as the [NeoRack](https://github.com/boazsegev/neorack) specification proposal, I would be happy to work with and see what features we would need to in order to help you implement an effective solution to your issue without the need to `hijack` the socket.

**Reply by trevorturk** [reply]
I don't have any practical advice (a lot of this stuff is above my pay grade 😬) but I did want to draw attention to an old comment from @byroot here: https://github.com/rails/rails/pull/44219#issuecomment-1051284194

> ...use a threaded server (or even just no threads at all) for the bread and butter transactional Rails workload, and a fiber server for your WebSocket stuffs or light proxying stuff.

Perhaps there's something here worth considering, if you can translate it from general advice into practical advice. My suspicion is that we might see multiple Ruby servers in use for complex workloads like this, but in my case currently I'm just naively running Falcon because the vast majority of my server time is spent waiting on upstream APIs (a very unusual case in the Ruby/Rails world, where I believe most people might reach for Node or Elixir etc...)
